AmplifyPlugins/DataStore/AWSDataStoreCategoryPlugin/Sync/SubscriptionSync/ReconcileAndLocalSave/ReconcileAndLocalSaveOperation.swift

```diff
@@ -137,186 +116,269 @@ class ReconcileAndLocalSaveOperation: AsynchronousOperation {
             return
         }
 
-        let localMetadata: MutationSyncMetadata?
-        do {
-            localMetadata = try storageAdapter.queryMutationSyncMetadata(for: remoteModel.model.id)
-        } catch {
-            stateMachine.notify(action: .errored(DataStoreError(error: error)))
+        guard !remoteModels.isEmpty else {
+            stateMachine.notify(action: .reconciled)
             return
         }
 
-        let queriedAction = Action.queried(remoteModel, localMetadata)
+        let remoteModelIds = remoteModels.map { remoteModel in
+            remoteModel.model.id
+        }
 
-        if log.logLevel == .debug {
-            log.debug("query local metadata: \(stopwatch.lap())s")
+        do {
+            try storageAdapter.transaction {
+                queryPendingMutations(forModelIds: remoteModelIds)
+                    .flatMap { mutationEvents -> Future<([RemoteModel], [LocalMetadata]), DataStoreError> in
```

Comment:
they're not at the same index because they may not be the same size, a remote model may not have local metadata. the tuple is representing: `(remoteModelsToApply, localMetadatas)` like "these are the remaining remote models we might need to, and here's the queried local metadatas (if exists) for those models"
```swift
queryPendingMutations(forModelIds: remoteModelIds) // get the pending mutations
                    .flatMap { mutationEvents -> Future<([RemoteModel], [LocalMetadata]), DataStoreError> in
                        let remoteModelsToApply = self.reconcile(remoteModels, pendingMutations: mutationEvents)
                        return self.queryLocalMetadata(remoteModelsToApply)
                    }
                    .flatMap { (remoteModelsToApply, localMetadatas) -> Future<Void, DataStoreError> in
```